python/aws_lambda_powertools/metrics/base.py

```diff
@@ -0,0 +1,210 @@
+with _schema_path.open() as f:
+    CLOUDWATCH_EMF_SCHEMA = json.load(f)
+
+
+validate_schema = fastjsonschema.compile(definition=CLOUDWATCH_EMF_SCHEMA)
+"""JSON Schema validator."""
+
+
+class MetricManager:
+    """Base class for metric functionality (namespace, metric, dimension, serialization)
+
+    MetricManager creates metrics asynchronously thanks to CloudWatch Embedded Metric Format (EMF).
+    CloudWatch EMF can create up to 100 metrics per EMF object
+    and metrics, dimensions, and namespace created via MetricManager
+    will adhere to the schema, will be serialized and validated against EMF Schema.
+
+    **Use `aws_lambda_powertools.metrics.metrics.Metrics` or
+    `aws_lambda_powertools.metrics.metric.single_metric` to create EMF metrics.**
+
+    Environment variables
+    ---------------------
+    POWERTOOLS_METRICS_NAMESPACE : str
+        metric namespace to be set for all metrics
+
+    Raises
+    ------
+    MetricUnitError
+        When metric metric isn't supported by CloudWatch
+    MetricValueError
+        When metric value isn't a number
+    UniqueNamespaceError
+        When an additional namespace is set
+    SchemaValidationError
+        When metric object fails EMF schema validation
+    """
+
+    def __init__(self, metric_set: Dict[str, str] = None, dimension_set: Dict = None, namespace: str = None):
+        self.metric_set = metric_set or {}
+        self.dimension_set = dimension_set or {}
+        self.namespace = os.getenv("POWERTOOLS_METRICS_NAMESPACE") or namespace
+
+    def add_namespace(self, name: str):
+        """Adds given metric namespace
+
+        Example
+        -------
+        **Add metric namespace**
+
+            metric.add_namespace(name="ServerlessAirline")
+
+        Parameters
+        ----------
+        name : str
+            Metric namespace
+        """
+        if self.namespace is not None:
+            raise UniqueNamespaceError(
+                f"Namespace '{self.namespace}' already set - Only one namespace is allowed across metrics"
+            )
+        logger.debug(f"Adding metrics namespace: {name}")
+        self.namespace = name
+
+    def add_metric(self, name: str, unit: MetricUnit, value: Union[float, int]):
+        """Adds given metric
+
+        Example
+        -------
+        **Add given metric using MetricUnit enum**
+
+            metric.add_metric(name="BookingConfirmation", unit=MetricUnit.Count, value=1)
+
+        **Add given metric using plain string as value unit**
+
+            metric.add_metric(name="BookingConfirmation", unit="Count", value=1)
+
+        Parameters
+        ----------
+        name : str
+            Metric name
+        unit : MetricUnit
+            `aws_lambda_powertools.helper.models.MetricUnit`
+        value : float
+            Metric value
+
+        Raises
+        ------
+        MetricUnitError
+            When metric unit is not supported by CloudWatch
+        """
+        if len(self.metric_set) == 100:
```

Comment:
The max is 100th metric so if exceed that by any chance we need to flush or else we lose all metrics

Reply:
I understand that, but you flush on the 101th metric getting added. I was wondering why you did this in the beginning of the method instead of the end. I am not sure if it matters much, but you are holding all 100 in memory until the next call instead of just flushing them on the 100th metric.